inc/Editor/DataBinding/Pagination.php

```diff
@@ -0,0 +1,212 @@
+		}
+
+		$query_id = $query->get_id();
+
+		// The query var value is an associative array with IDs as keys and
+		// values that are an associative array of input variables.
+		//
+		// We only expect a single key => value pair, but in the future we may
+		// decide to support more than one. This would allow us to control the
+		// pagination of multiple remote data blocks independently.
+		$untrusted_variables = $untrusted_variables[ $query_id ] ?? [];
+
+		if ( empty( $untrusted_variables ) || ! is_array( $untrusted_variables ) ) {
+			return [];
+		}
+
+		// Only accept pagination input variables that are defined in this query's
+		// input schema.
+		$input_schema = $query->get_input_schema();
+		$input_variables = [];
+		$pagination_input_variable_types = [
+			'ui:pagination_cursor',
+			'ui:pagination_cursor_next',
+			'ui:pagination_cursor_previous',
+			'ui:pagination_offset',
+			'ui:pagination_page',
+			'ui:pagination_per_page',
+		];
+
+		foreach ( $input_schema as $slug => $input ) {
+			if ( ! in_array( $input['type'] ?? null, $pagination_input_variable_types, true ) ) {
+				continue;
+			}
+
+			if ( ! array_key_exists( $slug, $untrusted_variables ) ) {
+				continue;
+			}
+
+			$value = $untrusted_variables[ $slug ];
+
+			if ( ! is_string( $value ) && ! is_int( $value ) ) {
+				continue;
+			}
+
+			$input_variables[ $slug ] = $value;
+		}
+
+		return $input_variables;
+	}
+
+	public static function format_pagination_data_for_query_response( array|null $pagination_data, array $query_input_schema, array $input_variables ): array {
+		// If the input schema is empty, it does not support pagination.
+		if ( empty( $query_input_schema ) ) {
+			return [];
+		}
+
+		// Find the appropriate pagination variables from the input schema.
+		$cursor_variable = null;
+		$cursor_next_variable = null;
+		$cursor_previous_variable = null;
+		$offset_variable = null;
+		$page_variable = null;
+		$per_page_variable = null;
+
+		foreach ( $query_input_schema as $slug => $input ) {
+			$type = $input['type'] ?? '';
+			if ( 'ui:pagination_cursor' === $type ) {
+				$cursor_variable = $slug;
+			} elseif ( 'ui:pagination_cursor_next' === $type ) {
+				$cursor_next_variable = $slug;
+			} elseif ( 'ui:pagination_cursor_previous' === $type ) {
+				$cursor_previous_variable = $slug;
+			} elseif ( 'ui:pagination_offset' === $type ) {
+				$offset_variable = $slug;
+			} elseif ( 'ui:pagination_page' === $type ) {
+				$page_variable = $slug;
+			} elseif ( 'ui:pagination_per_page' === $type ) {
+				$per_page_variable = $slug;
+			}
+		}
+
+		// Default pagination values
+		$current_per_page = $input_variables[ $per_page_variable ] ?? 10;
+		$pagination_input_variables = [];
+		$pagination_type = 'NONE';
+		$total_items = $pagination_data['total_items'] ?? null;
+
+		$pagination_input_variable_targets = [
+			'offset' => $offset_variable,
+			'page' => $page_variable,
+			'per_page' => $per_page_variable,
+		];
+
+		if ( $cursor_variable ) {
```

Comment:
Could this logic be documented just above this if statement, or as a function comment?

Reply:
Done in https://github.com/Automattic/remote-data-blocks/pull/449/commits/1465512f7006658a78925c02cf23b5a99e51b1d9